File: docs/DSA/neetcode150/src/go/permutation_in_string.go
package main

func checkInclusion(s1 string, s2 string) bool {
	counts := make(map[rune]int)
	for _, ch := range s1 {
		counts[ch]++
	}
	var l int
	for r, ch := range s2 {
		counts[ch]--
		for counts[ch] < 0 {
			counts[rune(s2[l])]++
			l++
		}
		if r-l+1 == len(s1) {
			return true
		}
	}
	return false
}

File: docs/DSA/src/py/permutation_in_string.py
from collections import Counter


class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        counts = Counter(s1)
        l = 0
        for r, ch in enumerate(s2):
            counts[ch] -= 1
            while counts[ch] < 0:
                counts[s2[l]] += 1
                l += 1

            if r - l + 1 == len(s1):
                return True
        return False
